rfcs/01391_async_sdk.md

```diff
@@ -0,0 +1,189 @@
+
+```rust
+oak::entrypoint!(oak_main => |_| {
+    // Grpc server setup ...
+
+    oak::run_event_loop(grpc_channel, handler);
+});
+
+async fn handler(requests: GrpcInvocations<HelloWorld>) {
+    // Initial setup
+    let translator = { /* ... */ };
+
+    requests.try_for_each_concurrent(
+        /* concurrency limit */ 2,
+        |request| async move {
+            match request {
+                HelloWorld::SayHello(request, response) => {
+                    let translation = translator.translate_async(TranslateRequest {
+                        text: request.greeting,
+                        from: "en".to_string(),
+                        to: "fr".to_string(),
+                    }).await?;
+                    response.send(HelloResponse {
+                        reply: format!("BONJOUR {}!", translated_greeting),
+                    })?;
+                }
+                // ...
+            }
+    }).await.expect("Error handling request");
+}
+```
+
+Proper client-side streaming ([an open issue](https://github.com/project-oak/oak/issues/97)) becomes straightforward now that the incoming requests can be represented as a [`Stream`](https://docs.rs/futures/0.3.5/futures/stream/trait.Stream.html):
+
+```rust
+match request {
+    // ...
+    HelloWorld::BidiHello(requests, responses) => {
+        requests.for_each(|request| {
+            response.send(HelloResponse {
+                reply: format!("HELLO {}!", request.greeting),
+            })?;
+        }).await;
+    }
+}
+```
+
+New functions defined on the existing [`Receiver`](https://project-oak.github.io/oak/sdk/doc/oak/io/struct.Receiver.html) allow asynchronous reads from channels:
+
+```rust
+impl<T: Decodable> Receiver<T> {
+    pub fn receive_async(&self) -> ChannelRead<T> { /* ... */ }
+    pub fn receive_stream(&self) -> ChannelReadStream<T> { /* ... */ }
+}
+```
+
+`ChannelRead` mentioned above is the asynchronous channel read primitive, implementing `Future<Output = Result<T, OakError>`. When `poll`ed, it notifies the executor that it is waiting for a particular channel to return data. The executor is in charge of checking for data on the channel, and waking up the future when the data is ready. 
+
+Since waiting on a channel is the only possible blocking operation (writes do not block in an Oak node), no other primitives are necessary. The `ChannelReadStream` for example is merely a repeated `ChannelRead` implementing `Stream<Item = Result<T, OakError>`.
+
+An `oak::spawn` method is provided that works exactly like [`tokio::spawn`](https://docs.rs/tokio/0.2.22/tokio/fn.spawn.html). When passed a future, a `JoinHandle` is returned that can be used to await the result of the spawned future. If this handle is dropped, the executor keeps polling the future to completion regardless, but there is no way to obtain the result of the computation. 
```

Comment:
Since we'll be building this on top of a `LocalPool`, adding this actually quite easy! 

But of course it can be a separate PR from the core executor etc 😉